src/natcap/invest/validation.py

```diff
@@ -998,55 +303,37 @@ def validate(args, spec, spatial_overlap_opts=None):
         # we don't need to try to validate them
         try:
             # Using deepcopy to make sure we don't modify the original spec
-            parameter_spec = copy.deepcopy(spec[key])
+            parameter_spec = copy.deepcopy(spec.get_input(key))
         except KeyError:
             LOGGER.debug(f'Provided key {key} does not exist in MODEL_SPEC')
             continue
 
-        param_type = parameter_spec['type']
-        if isinstance(param_type, set):
-            param_type = frozenset(param_type)
         # rewrite parameter_spec for any nested, conditional validity
-        axis_keys = None
-        if param_type == 'csv':
-            axis_keys = ['columns', 'rows']
-        elif param_type == 'vector' or 'vector' in param_type:
-            axis_keys = ['fields']
-        elif param_type == 'directory':
-            axis_keys = ['contents']
+        axis_keys = set(dir(parameter_spec)).intersection({'columns', 'rows', 'fields', 'contents'})
 
         if axis_keys:
             for axis_key in axis_keys:
-                if axis_key not in parameter_spec:
+                if getattr(parameter_spec, axis_key) is None:
                     continue
-                for nested_key, nested_spec in parameter_spec[axis_key].items():
-                    if ('required' in nested_spec
-                            and isinstance(nested_spec['required'], str)):
-                        parameter_spec[axis_key][nested_key]['required'] = (
+                for nested_spec in getattr(parameter_spec, axis_key):
+                    if (isinstance(nested_spec.required, str)):
+                        nested_spec.required = (
                             bool(_evaluate_expression(
-                                nested_spec['required'], expression_values)))
-
-        type_validation_func = _VALIDATION_FUNCS[param_type]
-
-        if type_validation_func is None:
-            # Validation for 'other' type must be performed by the user.
-            continue
```

Comment:
Huh, was this something from a leftover idea we once had?

Reply:
We used to support the "other" type but never used it. Now that we're using classes, if someone wanted to make another input type, they can subclass from `Input`!